ragthoven/executors/prompt_executor.py

```diff
@@ -30,33 +31,45 @@ def __init__(self, config: Config = None):
                 self.available_tools[tool_name] = get_class("ragthoven.tools", tool)()
 
     def get_tools(self, tools):
+        """
+        Build list of tool schemas for the model. Accepts either list of tool
+        names (resolved from self.available_tools) or pre-built json schemas.
+        """
         to_model_tools = None
         if tools is not None:
             to_model_tools = []
             for tool in tools:
+                # Already formatted schema
+                if isinstance(tool, dict) and tool.get("type") is not None:
+                    to_model_tools.append(tool)
+                    continue
+
                 tool_class: BaseFunCalling = self.available_tools[tool]
                 to_model_tools.append(tool_class.get_json_schema())
 
         return to_model_tools
 
-    def get_messages_prompt_results(self, messages, tools):
+    def get_messages_prompt_results(self, messages, tools, model=None):
         from litellm import completion
 
         to_model_tools = self.get_tools(tools)
         response = None
 
         try:
```

Comment:
The debug message `"What went wrong???"` could be more professional. Consider using `logger.error()` with a clearer message.